### ISSUE TITLE
Transparent retry should be per-hedge/retry

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#5075]
Transparent retry is limited to 1 attempt. It seems like that one attempt should be per-hedge and per-(higher-level) retry. In the current implementation transparent retry is limited to 1 attempt per RPC.

We should discuss whether the "scope" of the "1 attempt" and see if we could maybe even change the retry spec.

**Comment by dapengzhang0** [comment]
>It seems like that one attempt should be per-hedge and per-(higher-level) retry. 

I disagree with this because we can not tell if two hedging attempts are connecting to the same backend, in this case if the first hedging attempt receives a non-fatal status from the server's application logic, and after that the second hedging attempt fails with a "transparent-retryable" failure, then it does not make sense to do transparent retry for the second attempt because the backend's application logic has already run.

**Comment by ejona86** [comment]
> in this case if the first hedging attempt receives a non-fatal status from the server's application logic, and after that the second hedging attempt fails with a "transparent-retryable" failure, then it does not make sense to do transparent retry for the second attempt because the backend's application logic has already run.

That seems to be a normal case of "once we commit, kill all other streams."

**Comment by ejona86** [comment]
> in this case if the first hedging attempt receives a non-fatal status from the server's application logic, and after that the second hedging attempt fails with a "transparent-retryable" failure, then it does not make sense to do transparent retry for the second attempt because the backend's application logic has already run.

Also, that argument would apply today. There one only one transparent retry mentioned in your example.

**Comment by dapengzhang0** [comment]
There are two levels of commitment:
1. reached the application logic of the backend - no more transparent retry, but normal retry or hedging is allowed.
2. `committed` - no more retry or hedging allowed. Once we commit, kill all other streams.

I was talking about level 1.

For level 1. commitment, we need a global flag `noMoreTransparentRetry` minor all hedges base on my argument, regardless of if transparent retry is limited to 1 attempt per hedge. Given that we still need this global flag, what's the benefit of making the change?

**Comment by dapengzhang0** [comment]
>What is it protecting against?

I think it's as the spec says "This extra caution is needed because this case involves extra load on the wire". So we only allow at most (maxAttempts + 1) times of load on the wire. Otherwise, it may be doubled (maxAttempts * 2)

**Comment by dapengzhang0** [comment]
I think there was a debate on limit of transparent retries and then the spec got updated to the current version. + @markdroth @ericgribkoff
